jira/resilientsession.py

```diff
@@ -100,20 +104,137 @@ def __init__(self, timeout=None):
+        Raises:
+            Exception: Various exceptions as defined in py:method:`raise_on_error`.
+
+        Returns:
+            Response: The response.
+        """
+
+        retry_number = 0
+        exception: Optional[Exception] = None
+        response: Optional[Response] = None
+        response_or_exception: Optional[Union[ConnectionError, Response]]
+
+        processed_kwargs = self._jira_prepare(**_kwargs)
+
+        def is_retrying() -> bool:
+            """Helper method to say if we should still be retrying."""
+            return retry_number <= self.max_retries
+
+        while is_retrying():
+            response = None
+            exception = None
+
+            try:
+                response = super().request(method, url, **processed_kwargs)
+                if response.ok:
+                    self.__handle_known_ok_response_errors(response)
+                    return response
+            # Can catch further exceptions as required below
+            except ConnectionError as e:
+                exception = e
+
+            # Decide if we should keep retrying
+            response_or_exception = response if response is not None else exception
+            retry_number += 1
+            if is_retrying() and self.__recoverable(
+                response_or_exception, url, method.upper(), retry_number
+            ):
+                _prepare_retry_method.prepare(processed_kwargs)
+            else:
+                retry_number = self.max_retries + 1  # exit the while loop, as above max
+
+        if exception is not None:
+            # We got an exception we could not recover from
+            raise exception
+        elif raise_on_error(response, **processed_kwargs):
+            # raise_on_error will raise an exception if the response is invalid
+            return response
+        else:
+            # Shouldn't reach here...(but added for mypy's benefit)
+            raise RuntimeError("Expected a Response or Exception to raise!")
+
+    def __handle_known_ok_response_errors(self, response: Response):
+        """Responses that report ok may also have errors.
+
+        We can either log the error or raise the error as appropriate here.
+
+        Args:
+            response (Response): The response.
+        """
+        if not response.ok:
+            return  # We use self.__recoverable() to handle these
+        if (
+            len(response.content) == 0
+            and "X-Seraph-LoginReason" in response.headers
+            and "AUTHENTICATED_FAILED" in response.headers["X-Seraph-LoginReason"]
+        ):
+            LOG.warning("Atlassian's bug https://jira.atlassian.com/browse/JRA-41559")
+
     def __recoverable(
         self,
         response: Optional[Union[ConnectionError, Response]],
-        url: str,
-        request,
+        url: Union[str, bytes],
+        request_method: str,
         counter: int = 1,
     ):
+        """Return whether the request is recoverable and hence should be retried.
```

Comment:
what arguments would this second function take? Really this is just a bunch of logging and then two lines for the delay

Reply:
It would take `counter`, that is the number of retries performed so far, to calculate the delay. I don't feel strongly about this so feel free to keep the current structure, it's just a personal opinion about separating the two concerns.